cloud/azure/azure_rm_networkinterface.py

```diff
@@ -0,0 +1,619 @@
+        "type": "Microsoft.Network/networkInterfaces"
+    }
+'''
+
+from ansible.module_utils.basic import *
+from ansible.module_utils.azure_rm_common import *
+
+try:
+    from msrestazure.azure_exceptions import CloudError
+    from azure.common import AzureMissingResourceHttpError
+    from azure.mgmt.network.models import NetworkInterface, NetworkInterfaceIPConfiguration, Subnet, \
+                                          PublicIPAddress, NetworkSecurityGroup
+    from azure.mgmt.network.models.network_management_client_enums import IPAllocationMethod
+except ImportError:
+    # This is handled in azure_rm_common
+    pass
+
+
+NAME_PATTERN = re.compile(r"^[a-z][a-z0-9-]{1,61}[a-z0-9]$")
+
+
+def nic_to_dict(nic):
+    result = dict(
+        id=nic.id,
+        name=nic.name,
+        type=nic.type,
+        location=nic.location,
+        tags=nic.tags,
+        network_security_group=dict(),
+        ip_configuration=dict(
+            name=nic.ip_configurations[0].name,
+            private_ip_address=nic.ip_configurations[0].private_ip_address,
+            private_ip_allocation_method=nic.ip_configurations[0].private_ip_allocation_method.value,
+            subnet=dict(),
+            public_ip_address=dict(),
+        ),
+        dns_settings=dict(
+            dns_servers=nic.dns_settings.dns_servers,
+            applied_dns_servers=nic.dns_settings.applied_dns_servers,
+            internal_dns_name_label=nic.dns_settings.internal_dns_name_label,
+            internal_fqdn=nic.dns_settings.internal_fqdn
+        ),
+        mac_address=nic.mac_address,
+        primary=nic.primary,
+        enable_ip_forwarding=nic.enable_ip_forwarding,
+        provisioning_state=nic.provisioning_state,
+        etag=nic.etag,
+    )
+
+    if nic.network_security_group:
+        result['network_security_group']['id'] = nic.network_security_group.id
+        id_keys = azure_id_to_dict(nic.network_security_group.id)
+        result['network_security_group']['name'] = id_keys['networkSecurityGroups']
+
+    if nic.ip_configurations[0].subnet:
+        result['ip_configuration']['subnet']['id'] = \
+            nic.ip_configurations[0].subnet.id
+        id_keys = azure_id_to_dict(nic.ip_configurations[0].subnet.id)
+        result['ip_configuration']['subnet']['virtual_network_name'] = id_keys['virtualNetworks']
+        result['ip_configuration']['subnet']['name'] = id_keys['subnets']
+
+    if nic.ip_configurations[0].public_ip_address:
+        result['ip_configuration']['public_ip_address']['id'] = \
+            nic.ip_configurations[0].public_ip_address.id
+        id_keys = azure_id_to_dict(nic.ip_configurations[0].public_ip_address.id)
+        result['ip_configuration']['public_ip_address']['name'] = id_keys['publicIPAddresses']
+
+    return result
+
+
+class AzureRMNetworkInterface(AzureRMModuleBase):
+
+    def __init__(self):
+
+        self.module_arg_spec = dict(
+            resource_group=dict(required=True),
```

Comment:
you should use type='str', since in 2.1 we will need all types to be explicit unless you want 'raw' functionality (type is whatever is supplied), same for name and other's below